### PR TITLE
audio: Add a NetBSD audio output backend

### PR DESCRIPTION
**Pull request by niacat**
The OSS backend is problematic on NetBSD, requiring e.g. --autosync=30 to avoid frame dropping. This one generally works better.

Documentation for the native interface:
https://netbsd.gw.com/cgi-bin/man-cgi?audio+4

### CODE REVIEW

**Comment by niacat** [comment]
CI is failing on a timeout:

Error response from daemon: Get https://registry.cirno.systems/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)

**Comment by wm4** [comment]
That's another audio API that exists only on an obscure OS with even smaller userbase which cares about audio, can never be ported or tested on other OSes, and is essentially unmaintainable.

Look at what happened to the OpenBSD equivalent to it (ao_sndio). It's half-broken, and I don't even know if anyone uses it.

Why not just write an ALSA backend for NetBSD?

**Comment by niacat** [comment]
I'm willing to maintain this. I recently submitted some code to Firefox using this API and got it merged, and I've also been actively maintaining the SDL2 audio code.

I'm vaguely aware of who wrote the sndio backend and I'm sure they wrote it for sndio on Linux (sndio supports multiple platforms). It's possible that OpenBSD's emulation of OSS is working well enough with mpv that nobody cared.

As for ALSA, it needs to be configured to be useful on non-Linux, they don't care about us much, and I'm trying to avoid layering bugs that result from pushing audio through third-party APIs.

We have a quite popular mpv package and people do notice when it breaks, which is what drove me to write this in the first place.

**Comment by wm4** [comment]
It surprises me to hear that Firefox accepted a backend that's helping maybe a few hundred people at most, while they actually removed the ALSA backend, forcing several magnitude more Linux users to use use half-broken Pöttering-abandon-ware.

It also makes no sense to add new netbsd backends to dozens of software projects (that now all need Netbsd-specific maintenance), while adding such a backend to ALSA would automatically support all these projects at once (except Firefox), and would require only maintaining the ALSA code, instead of N projects.

Your argument about layers doesn't hold up. What difference is there between having the additionally needed layer in the AO code, or as ALSA backend? Except that the latter would be put between two stable and widely used APIs, instead of some obscure project that changes its internal APIs every other month.

**Comment by iamleot** [comment]
I don't think that NetBSD is an _obscure OS_ or if that a backend that's _helping maybe a few hundred people at most_ are good technical reasons to not merge this PR.

Also, since @niacat commit's https://github.com/NetBSD/pkgsrc/commit/bc5af1fe8c6b820bf31d0c1abcd84bb5a70fda53 `netbsd` ao is the preferred audio output on NetBSD so if there would be a problem in NetBSD with it, it would be probably the one noticed first.

**Comment by wm4** [comment]
> I don't think that NetBSD is an obscure OS

It definitely is for multimedia.

> Also, since @niacat commit's NetBSD/pkgsrc@bc5af1f netbsd ao is the preferred audio output on NetBSD so if there would be a problem in NetBSD with it, it would be probably the one noticed first.

Unless you're OK with the AO randomly bitrotting and you having to fix it again,  the existence of the AO is at first _my_ problem, because I have to reason about what weird behavior the AO might have, which will matter in corner cases. The audio "state machine" that deals with the AOs is already ridiculously complex.

**Comment by iamleot** [comment]
> Unless you're OK with the AO randomly bitrotting and you having to fix it again, the existence of the AO is at first _my_ problem, because I have to reason about what weird behavior the AO might have, which will matter in corner cases. The audio "state machine" that deals with the AOs is already ridiculously complex.

@niacat offered to maintain it and if you do any change I'm pretty sure that we can help to test/adjust it if needed by just contacting us.

**Comment by wm4** [comment]
Again, I don't want to have to think about another audio API being used in mpv. Nor waiting for someone when I make a change to the internal audio API in mpv.

And again I have to ask: why did NetBSD invent its own audio API? Why not improve OSS? Or discouraging application authors from using the kernel interface, and providing an ALSA backend instead?

**Comment by polygonalenippel** [comment]
As a bystander reading this issue, I'd like to provide my point of view.
niacat seeks to have his PR merged for his niche OS, but you (@wm4) refuse on the grounds that
* The audio subsystem of his OS is inherently flawed
* It adds extra complexity to the already complex audio handling of mpv

Both these points result in that they add extra work for you that you would rather not want to deal with. However, these people (niacat, iamleot) offer to take that burden off of you by taking responsibility for this part.

Why do you still insist that this PR would not be merged? Is it mainly ideological reasons? Or some other technical aspect which I am not seeing?
Just want to make sure you properly outweigh your decisions for the long run. Also wanted to provide my 2ct for this great software. Keep up the good work, guys!

**Comment by wm4** [comment]
> The audio subsystem of his OS is inherently flawed

Maybe, I don't know yet, but most audio APIs have annoying flaws and you get to deal with the union of them. Sndio certainly seemed to have some major flaws.

> Why do you still insist that this PR would not be merged?

If I absolutely did, I'd have closed this PR long ago. However, I want some major concessions at least and good answers.

**Comment by niacat** [comment]
>And again I have to ask: why did NetBSD invent its own audio API? Why not improve OSS? Or discouraging application authors from using the kernel interface, and providing an ALSA backend instead?

It's the SunOS/Solaris API, we didn't invent it. Some NetBSD extensions are being used that are necessary because mpv requires accurate delay measurement, so this code is currently NetBSD-only.

The API has been stable for more than 20 years, see the code in projects such as mpg123 and libao. From my perspective, it's cleaner and simpler to use than OSS.

**Comment by wm4** [comment]
MPlayer still has the Sun API: https://github.com/pigoz/mplayer-svn/blob/master/libao2/ao_sun.c

I find nothing clean about ioctls, though.

**Comment by niacat** [comment]
>I find nothing clean about ioctls, though.

Better remove the OSS code, then.

**Comment by wm4** [comment]
I'd love to remove that, but FreeBSD people would protest too much.

**Comment by wm4** [comment]
I'll leave it to other mpv members to decide whether this should go in.
If there's a positive response, I'll actually make a review of the code.

**Comment by Akemi** [comment]
i am indifferent. though at least someone should feel responsible for this part of the code and also directly related parts in the core, if problems should arise. if it makes problems in the future, none is fixing it in a timely fashion, or it is majorly in the way of other parts it should be removed.

**Comment by avih** [comment]
I'm also OK with it, considering the code is relatively small and OP suggests to maintain it.

But I'm also suggesting a new practice:

Whenever the core changes in a way which is not compatible with modules which have a maintainer - the module will get disabled until the maintainer steps in and does the necessary adjustments.

I think this should make changes at the core easier to do.

**Comment by kevmitch** [comment]
Improved HURD support? Why haven't we merged this already?

Seriously though. seems fine to me since @niacat is agreeing to maintain it. Code looks pretty simple and keeps to itself.

File: audio/out/ao_netbsd.c [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+    AUDIO_INITINFO(&info);
+    pinfo = &info.play;
+
+    if (ao->device != NULL) {
+         (void)snprintf(device, sizeof(device), "/dev/audio%s",
+                        ao->device + sizeof("netbsd/") - 1);
+    }
+
+    MP_ERR(ao, "Opening device %s", ao->device);
```
Comment (haasn) [inline audio/out/ao_netbsd.c]: I guess this was meant to be `MP_VERBOSE`.

**Comment by haasn** [comment]
I haven't done an in-depth review but it looks fine to me overall. I'm in favor of merging (in general, not necessarily as-is):

- The code is useful to someone
- This someone has both written the code and agreed to maintain in
- The code is conditionally compiled only on relevant platforms
- The code is isolated and doesn't introduce complexity in the rest of mpv that wasn't already there
- The code is relatively small and not confusing to read
- The NetBSD API is unlikely to change; so the only breakage potential would be a change of the AO API, which due to its simplicity is unlikely to affect this AO in a way we can't trivially adapt ourselves even without knowledge of NetBSD.

My main concern would be making sure the CI compiles this code, because it's such an obscure platform. Do we either have the ability to test on NetBSD, or alternatively can we compile the code on non-NetBSD platforms with the right headers?

**Comment by Dudemanguy** [comment]
I think it's fine. I'm not familiar with the audio stuff, but the code look simple and understandable enough to me. Even if it completely bitrots one day, there's always git rm. The CI thing pointed out earlier is more of a concern imo.

**Comment by wm4** [comment]
Meh, fine, I'll make a review.

> My main concern would be making sure the CI compiles this code, because it's such an obscure platform.

That sounds like a lot of work. I don't know if you can even cross-compile to BSD. And you'd need to do it for ffmpeg etc. too.

File: audio/out/ao_netbsd.c [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
```
Comment (wm4) [inline audio/out/ao_netbsd.c]: I think the point 2 is too strict. Can it be 0-clause BSD or ISC license?

File: audio/out/ao_netbsd.c [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+#include <errno.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "options/m_option.h"
+#include "common/msg.h"
+
+#include "audio/format.h"
+#include "ao.h"
+#include "internal.h"
```
Comment (wm4) [inline audio/out/ao_netbsd.c]: New code should sort includes alphabetically, see https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md#header-include-statement-order

File: audio/out/ao_netbsd.c [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+#define NETBSD_MAX_CHANNELS (12)
+#endif
+
+#ifndef NETBSD_MIN_RATE
+#define NETBSD_MIN_RATE (1000)
+#endif
+
+#ifndef NETBSD_MAX_RATE
+#define NETBSD_MAX_RATE (192000)
+#endif
```
Comment (wm4) [inline audio/out/ao_netbsd.c]: What are these defines? Remove them.

File: audio/out/ao_netbsd.c [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+    struct audio_info info;
+    struct audio_prinfo *pinfo;
+    struct mp_chmap_sel sel = {0};
+
+    AUDIO_INITINFO(&info);
+    pinfo = &info.play;
+
+    if (ao->device != NULL) {
+         (void)snprintf(device, sizeof(device), "/dev/audio%s",
+                        ao->device + sizeof("netbsd/") - 1);
```
Comment (wm4) [inline audio/out/ao_netbsd.c]: Don't use the casting to void idiom. It's useful sometimes, but here it's just noise.

I don't think what you try to do with ao->device works. At all.

File: audio/out/ao_netbsd.c [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+    }
+
+    MP_ERR(ao, "Opening device %s\n", device);
+    if ((p->fd = open(device, O_WRONLY)) == -1) {
```
Comment (wm4) [inline audio/out/ao_netbsd.c]: Don't assign within the if condition.

File: audio/out/ao_netbsd.c [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+               device, mp_strerror(errno));
+        goto fail;
+    }
+
+    info.mode = AUMODE_PLAY;
+
+    for (int n = 1; n <= NETBSD_MAX_CHANNELS; n++) {
+        struct mp_chmap map;
+
+        mp_chmap_from_channels(&map, n);
```
Comment (wm4) [inline audio/out/ao_netbsd.c]: I bet this is wrong and random and untested.

File: audio/out/ao_netbsd.c [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+    struct priv *p = ao->priv;
+    struct audio_info info;
+    struct audio_offset offset;
+
+    if (ioctl(p->fd, AUDIO_GETINFO, &info) == -1) {
+        MP_ERR(ao, "AUDIO_GETINFO failed: %s\n", mp_strerror(errno));
+        return;
+    }
+
+    (void)ioctl(p->fd, AUDIO_DRAIN, NULL);
```
Comment (wm4) [inline audio/out/ao_netbsd.c]: Apparently that waits until the buffer plays. It's supposed to do the opposite. And we want (almost) never to block on the audio device either (in reset() we certainly do not).

File: audio/out/ao_netbsd.c [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+    (void)ioctl(p->fd, AUDIO_GETOOFFS, &offset); /* reset deltablks */
+    p->total_blocks = p->total_bytes / info.blocksize;
+}
+
+static int get_space(struct ao *ao)
+{
+    return NETBSD_BUF_SIZE;
```
Comment (wm4) [inline audio/out/ao_netbsd.c]: This is unacceptable. It must return the number of samples ``play()``` will accept.

File: audio/out/ao_netbsd.c [hunk 10]
```diff
@@ -0,0 +1,280 @@
...
+static int play(struct ao *ao, void **data, int samples, int flags)
+{
+    struct priv *p = ao->priv;
+    int len = samples * ao->sstride;
+    int ret;
+
+    if (len == 0)
+        return 0;
+
+    errno = 0;
```
Comment (wm4) [inline audio/out/ao_netbsd.c]: Why clear errno?

**Comment by iamleot** [comment]
It seems that there were on going discussion, reviews and patch in progress.
It also seems that several members of mpv-player were okay to possibly merge this pull request.

Was it accidentally closed?

**Comment by Akemi** [comment]
it was closed because of the hostility in #7643

**Comment by iamleot** [comment]
I think that closing it without any rationale/discussion is similarly hostile as well though. :(

**Comment by CounterPillow** [comment]
nah it's just what you get for being a shitter

**Comment by Akemi** [comment]
the rationale was given by me in form of that other PR. apparently no discussion was wanted by the other side, seen by that action, and we are not going to discuss something in that kind of form.

this is far from hostile, but just an easy and obvious decision. one of the requirements for this PR being merged was that the OP takes responsibility for these changes. this means we have to work together in a somewhat civilised way. with those actions any kind of civilised interaction is gone and we simply decided to not work together in that way.

**Comment by iamleot** [comment]
I think that #7643 was just a provocation of reasoning and attitude behind the removal of `ao_oss` (<https://github.com/mpv-player/mpv/commit/bca917f6d2c50d9df3a4140b7be7beb6d567aa5b>) and `ao_sndio` (<https://github.com/mpv-player/mpv/commit/71d218eae4b4d93ada34ff74906f71ad359c84bc>).

In free software, at least in a "not-one-man-project", I would expect that before removing possible functionality that could be still used by the community someone come up with a possible proposal regarding removing it and *then* if the functionality is no longer used or there are possible alternatives to it, it's removed.

**Comment by Akemi** [comment]
it doesn't really matter, and i am sure it is related, such behaviour is just not acceptable.

yeah it could have been explicitly deprecated before removal, which i am not sure it might have been partially already.

there are alternatives apparently.

if there is none to maintain it, it will eventually be removed. in free software, usually people step in if they want to keep something and start maintaining it.

**Comment by wm4** [comment]
sndio was disabled by default for years.
